gnovm/pkg/packages/pkglist.go

```diff
@@ -0,0 +1,110 @@
+package packages
+
+import (
+	"errors"
+	"fmt"
+)
+
+type (
+	PkgList       []*Package
+	SortedPkgList []*Package
+)
+
+func (pl PkgList) Get(pkgPath string) *Package {
+	for _, p := range pl {
+		if p.ImportPath == pkgPath {
+			return p
+		}
+	}
+	return nil
+}
+
+var ErrPackageNotFound = errors.New("package not found")
+
+func (pl PkgList) GetByDir(dir string) *Package {
+	for _, p := range pl {
+		if p.Dir == dir {
+			return p
+		}
+	}
+	return nil
+}
+
+// sortPkgs sorts the given packages by their dependencies.
+func (pl PkgList) Sort() (SortedPkgList, error) {
+	visited := make(map[string]bool)
+	onStack := make(map[string]bool)
+	sortedPkgs := make([]*Package, 0, len(pl))
+
+	// Visit all packages
+	for _, p := range pl {
+		if err := visitPackage(p, pl, visited, onStack, &sortedPkgs); err != nil {
+			return nil, err
+		}
+	}
+
+	return sortedPkgs, nil
+}
```

Comment:
this function looks like searching would be possible in constant time, if you build the package map before the loop.